### DISCUSSION TITLE
Performance margins on urequests

### DISCUSSION THREAD

**Discussion by the-maldridge** [micropython/micropython#10631] [General]
I'm currently building a prototype that fetches a packet of control information from a webserver in a very tight loop.  The webserver runs on a regular amd64 server and is written in Go.  The results from `hey` suggests that my P99 latency is well under one millisecond, which seems reasonable for the test endpoint that just serves the time.

On the micropython side I am testing with an rpi pico w, and polling the `/time` endpoint of my server in a tight loop.  I notice, however, that I only get about 2 requests per second and this seems very slow.  Is this the expected performance that I should be able to get out of micropython?  I suspect this code may need to be rewritten in C to achieve the performance margins I require, but I'd appreciate any input before I go that route.

**Comment by kjm1102** [comment]
urequests does a dns lookup every time it runs so if you're calling your server by name rather than ip address that could be where some of the delay is? Also https (port 443) takes longer than http (port 80) because of the encryption.

**Reply by the-maldridge** [reply]
Hmm, I am accessing the server unencrypted over port 80 by a local IP, so I'm assuming that this should be as fast as it could possibly be.

**Comment by kjm1102** [comment]
If I'm in a tight control loop I usually dispense with urequests & use a function like this
```
def http(url):
  print("get(", url, ")"); _, _, host, path = url.split('/', 3); n=0
  print('host', host, '  path', path)
  t1=time.ticks_ms(); addr=socket.getaddrinfo(host, 80)[0][-1]; t=time.ticks_ms()-t1; print('dns in', t/1000, addr, 'from', socket.dnsserver())
  t1=time.ticks_ms(); s=socket.socket(); s.connect(addr); s.write(b'GET /%s HTTP/1.0\r\nHost: %s\r\n\r\n' % (path, host))
  while True:
    data = s.recv(900); t2=time.ticks_ms(); t=t2-t1; t1=t2; n+=1
    if data: print('    srvr reply', str(n), ' in', str(t/1000)); print(str(data, 'utf8'))
    else: break
  s.close()
```
